### PR TITLE
Add single-page deploy

### PR DESCRIPTION
**Pull request by harmenjanssen**
This is a Work in Progress used to discuss the concept.

### CODE REVIEW

File: includes/Archiver.php [hunk 1]
```diff
@@ -22,11 +22,16 @@ class Archiver
     private $tasks = [];
     private $files = [];
     private $urls = [];
+    private $excludedUrls = [];
 
-    public function __construct()
+    public function __construct(?int $post_id = null)
```
Comment (HammenWS) [inline includes/Archiver.php:27]: Does this match with our requirements?

File: includes/Archiver.php [hunk 2]
```diff
@@ -41,6 +46,9 @@ public function __construct()
         $this->urls = Utils::option_string_to_array(
             Simply_Static\Options::instance()->get('additional_urls')
         );
+        $this->excludedUrls = Utils::option_string_to_array(
+            Simply_Static\Options::instance()->get('urls_to_exclude')
```
Comment (HammenWS) [inline includes/Archiver.php]: Doublecheck if this is the way to transform the data.

File: includes/Tasks/SetupSingleTask.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                [
+                    'url' => $url,
+                    'do_not_save' => '0',
+                    'do_not_follow' => '1',
+                ],
+            ]
+        );
+        Options::instance()->set(
+            'urls_to_exclude',
+            Utils::array_to_option_string($excludedUrls)
```
Comment (HammenWS) [inline includes/Tasks/SetupSingleTask.php]: Check if this is the way to transform the data

File: includes/Api.php [hunk 4]
```diff
@@ -37,7 +37,8 @@ public function register_api_endpoints()
                     'methods' => 'POST',
                     'callback' => [$this, $callback],
                     'permission_callback' => function () {
-                        return current_user_can('edit_posts');
+                        return true;
```
Comment (HammenWS) [inline includes/Api.php:40]: This should be turned back.

File: includes/Archiver.php (hunk 1 above)
Comment (HammenWS) [reply includes/Archiver.php]: It does

File: includes/Archiver.php (hunk 2 above)
Comment (HammenWS) [reply includes/Archiver.php]: Checked

File: includes/Tasks/SetupSingleTask.php (hunk 3 above)
Comment (HammenWS) [reply includes/Tasks/SetupSingleTask.php]: Checked

File: includes/StaticDeployJob.php [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+                'store_initial_options',
+                'setup',
+                'fetch_urls',
+                'transfer_files_locally',
+                'wrapup',
+                'restore_initial_options',
+                'sync',
+                'invalidate',
+            ];
+        }
```
Comment (harmenjanssen) [inline includes/StaticDeployJob.php:333]: Do you think it's more clear in this function to only replace the `setup` task with a ternary statement? It saves repeating the entire task list.

File: includes/StaticDeployJob.php [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+                \RecursiveDirectoryIterator::SKIP_DOTS
+            ),
+            \RecursiveIteratorIterator::CHILD_FIRST
+        );
+        foreach ($files as $fileinfo) {
+            $todo = $fileinfo->isDir() ? 'rmdir' : 'unlink';
+            $todo($fileinfo->getRealPath());
+        }
+    }
+}
```
Comment (harmenjanssen) [inline includes/StaticDeployJob.php:358]: Could you reorder this class's functions so public functions are on top?

File: includes/Tasks/InvalidateTask.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+use Simply_Static\Task;
+use Simply_Static\Util;
+
+
+class InvalidateTask extends Task {
+
+    protected static $task_name = 'invalidate';
+
+    public function perform() {
+        Util::debug_log('Invalidate the shizzle');
```
Comment (harmenjanssen) [inline includes/Tasks/InvalidateTask.php:17]:
```suggestion
        Util::debug_log('Invalidate the Cloudfront distribution.');
```

File: includes/Tasks/SyncTask.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+use Grrr\SimplyStaticDeploy\Syncer;
+use Simply_Static\Task;
+use Simply_Static\Util;
+
+class SyncTask extends Task {
+
+    protected static $task_name = 'sync';
+
+    public function perform() {
+        Util::debug_log('Sync shizzle to s3');
```
Comment (harmenjanssen) [inline includes/Tasks/SyncTask.php:17]:
```suggestion
        Util::debug_log('Sync site to s3');
```

File: includes/Tasks/StoreInitialOptionsTask.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    protected static $task_name = 'store_initial_options';
+
+    // const CLEAR_FILTER = 'simply_static_deploy_clear_directory';
+    const FILES_FILTER = 'simply_static_deploy_additional_files';
+    // const PHP_FILTER = 'simply_static_deploy_php_execution_time';
+    const URLS_FILTER = 'simply_static_deploy_additional_urls';
+
+    // const MODIFY_ACTION = 'simply_static_deploy_modify_generated_files';
+
```
Comment (harmenjanssen) [inline includes/Tasks/StoreInitialOptionsTask.php:22]: Are the commented out constants still necessary?

**Review by harmenjanssen — COMMENTED** [review]
@HammenWS This looks good! 
I have some cosmetic comments.

We should also decide what to do with Prettier: I see you reverted some of Prettier's changes. If we want to use it we have to go all the way. If not, we should drop the first commit from this PR.

File: includes/Api.php [hunk 10]
```diff
@@ -1,76 +1,108 @@
...
 
+    protected $staticDeployJob;
+
     public function __construct(Config $config) {
         $this->config = $config;
+        $this->staticDeployJob = new StaticDeployJob;
     }
 
     public function register() {
         add_action('rest_api_init', [$this, 'register_api_endpoints']);
```
[the remark this answers is not in the document]
Comment (HammenWS) [reply includes/Api.php:34]: But the `rest_api_init` is an action from WordPress and will never change.

[the remark this answers is not in the document]
Comment (harmenjanssen) [reply includes/Api.php:36]: Gotcha, fair enough.

File: CHANGELOG.md [hunk 11]
```diff
@@ -3,6 +3,24 @@
...
 ## v1.0.0 (2020-04-16)
+
 The plugin version was bumped from `v0.2.x` to `v1.x.x` to improve updating via Composer, due to restrictions in the [caret version range](https://getcomposer.org/doc/articles/versions.md#caret-version-range-). From the docs:
 
 > For pre-1.0 versions it also acts with safety in mind and treats `^0.3` as `>=0.3.0 <0.4.0`.
+
+## v2.0.0 (2020-10-19)
+
+From version v2.0.0 each deploy task will be done in the background.
+This means your browser window won't have to be always open while deploying. And all the tasks will seperate requests, so the idle timeout limit of the server won't be reached.
```
Comment (harmenjanssen) [inline CHANGELOG.md:14]:
```suggestion
This means your browser window won't have to stay open while deploying. All the tasks will make separate requests, so the idle timeout limit of the server won't be reached.
```

**Review by harmenjanssen — COMMENTED** [review]
I've tweaked one single sentence!

**Pull request merged**